xpti/doc/SYCL_Tracing_Implementation.md

```diff
@@ -0,0 +1,86 @@
+
+  To instrument this part of the code, the ```Command::addDep``` methods of the Command object are instrumented to create the trace points and notify all subscribers. 
+    
+This concludes all the changes that were made to the SYCL runtime to support tracing. The next section talks about the XPTI framework that allows applications and runtimes to efficiently capture, record and emit trace notifications for important events during the run.
+
+# SYCL Tracing - Part II
+The architecture of the XPTI is discussed in other documents, but at the rudimentary level, the instrumentation in any runtime or application links to a static stub library that implements all the functions in the library. If the tracing is enabled through the environment variable ```XPTI_TRACE_ENABLE=1```, the stub library checks to see if the framework dispatcher is registered. This dispatcher is registered through an environment variable ```XPTI_FRAMEWORK_DISPATCHER=/path/to/libxptifw.so```. If tracing is turned on and the dynamic loading of the framework dispatcher is successful, then the proxy library creates a dipatch table for all the trampoline functions used in the instrumentation of the application or runtime. 
+
+The static library is the only dependency for building the SYCL runtime and is currently available under ```llvm/xpti```. In the current organization at the file system level, the API specification for the instrumentation framework is available with the static library under ```llvm/xpti/include```. The framework is divided into two parts: (1) the implementation of the proxy/stub library that will be compiled to create a static library to be linked with SYCL runtime and (2) a dynamic library with can be registered as the framework dipatcher and is not required by the SYCL runtime that is not a part of the ```llvm``` project. The dynamic library depends on the API specification that is a part of the static proxy library. Using the specification in the ```llvm/xpti```, an implementation of the dynamic libray can be built.
+
+## The Proxy library
+The proxy library implements all of the public functions that are a part of the XPTI tracing infrastructure. Each function however is a stub that checks to see if tracing has been enabled. If so, it forwards the call to the same function in the framework dispatcher or the dynamic component of this framework.
+
+   ```c++
+   XPTI_EXPORT_API void xptiInitialize(const char *stream,
+      uint32_t maj,
+      uint32_t min,
+      const char *version) {
+      // Static object g_loader will return true if
+      // XPTI_TRACE_ENABLE=1 and the dynamic dispatcher
+      // library in XPTI_FRAMEWORK_DISPATCHER is valid
+      // and successfully loaded.
+      //
+      if (xpti::g_loader.noErrors()) {
+          void *f = xpti::g_loader.functionByIndex(XPTI_INITIALIZE);
+          if (f) {
+              (*(xpti_initialize_t)f)(stream, maj, min, version);

```

Comment:
And if f is false?

Reply:
It will fall through to the appropriate error value that should be returned. It is conceivable that someone might partially implement the API, hence the check.

Reply:
There are a couple of APIs that are void. While not knowing if the dispatcher is valid or not is okay for these functions, it may make sense to return an error code so it can be handled earlier by the calling code.